File: 2021/python/1657.determin-two-string-close.py
"""
1657. Determine if Two Strings Are Close

Two strings are considered close if you can attain one from the other using the following operations:

Operation 1: Swap any two existing characters.
For example, abcde -> aecdb
Operation 2: Transform every occurrence of one existing character into another existing character, and do the same with the other character.
For example, aacabb -> bbcbaa (all a's turn into b's, and all b's turn into a's)
You can use the operations on either string as many times as necessary.

Given two strings, word1 and word2, return true if word1 and word2 are close, and false otherwise.



Example 1:

Input: word1 = "abc", word2 = "bca"
Output: true
Explanation: You can attain word2 from word1 in 2 operations.
Apply Operation 1: "abc" -> "acb"
Apply Operation 1: "acb" -> "bca"
Example 2:

Input: word1 = "a", word2 = "aa"
Output: false
Explanation: It is impossible to attain word2 from word1, or vice versa, in any number of operations.
Example 3:

Input: word1 = "cabbba", word2 = "abbccc"
Output: true
Explanation: You can attain word2 from word1 in 3 operations.
Apply Operation 1: "cabbba" -> "caabbb"
Apply Operation 2: "caabbb" -> "baaccc"
Apply Operation 2: "baaccc" -> "abbccc"
Example 4:

Input: word1 = "cabbba", word2 = "aabbss"
Output: false
Explanation: It is impossible to attain word2 from word1, or vice versa, in any amount of operations.


Constraints:

1 <= word1.length, word2.length <= 105
word1 and word2 contain only lowercase English letters.
"""

def closeStrings(word1: str, word2: str) -> bool:
    # if not the same length, not the same pattern
    if len(word1) != len(word2):
        return False
    # count character frequency
    count1 = {ch : word1.count(ch) for ch in set(word1)}
    count2 = {}
    # count word2 frequency and check whether character in word1, if not, terminate early
    for key in word2:
        if key not in count1:
            return False
        count2[key] = count2.get(key, 0) + 1
    # sort dict by values (frequency)
    count1 = sorted(count1.values())
    count2 = sorted(count2.values())
    # check values are equal, not equal, terminate early
    for v1, v2 in zip(count1, count2:
        if v1 != v2:
            return False
    return True

File: 2021/java/1657.determin-two-string-close.java
class Solution {
    public boolean isCloseStrings(String word1, String word2) {
        if (word1 == null) return word2 == null;
        if (word2 == null) return word2 == null;
        if (word1.length() != word2.length()) return false;
        // word ony contains lower case 26 letters
        int[] count1 = new int[26];
        int[] count2 = new int[26];
        for (char ch : word1.toCharArray) {
            count1[ch - 'a']++;
        }
        for (char ch : word2.toCharArray) {
            // if not in word1, not the same pattern, terminate early
            if (count1[ch - 'a'] < 1) return false;
            count2[ch - 'a']++;
        }
        Arrays.sort(count1);
        Arrays.sort(count2);
        for (int i = 0; i < 26; i++) {
            // not the same pattern, terminate ealry
            if (count1[i] != count2[i]) return false;
        }
        return true;
    }
}
